infrastructure-userdb/display-name.rkt

```diff
@@ -0,0 +1,163 @@
+    email->display-name))
+
+(define (email->plain-tag email)
+  (string->symbol (string-append "author:" email)))
+
+(define (make-display-name email [plain-tag (email->plain-tag email)])
+  (match-define (list local-part)
+    (regexp-match #rx"^[^@]*" email))
+  (define short-hash
+    (substring (bytes->hex-string
+                (sha256-bytes (open-input-string email)))
+               0
+               7)) ; on the very scientific basis that this is what Git uses
+  (display-name
+   (string->immutable-string email)
+   (string->immutable-string local-part)
+   (string->immutable-string short-hash)
+   plain-tag
+   (string->symbol (string-append "author:" local-part "λ" short-hash))))
+
+(module+ test
+  (for ([expected (list (display-name "philip@philipmcgrath.com"
+                                      "philip"
+                                      "9411372"
+                                      'author:philip@philipmcgrath.com
+                                      'author:philipλ9411372)
+                        (display-name "samth@racket-lang.org"
+                                      "samth"
+                                      "5dd761c"
+                                      'author:samth@racket-lang.org
+                                      'author:samthλ5dd761c))])
+    (define actual
+      (email->display-name (display-name-email expected)))
+    (unless (equal? actual expected)
+      (raise-arguments-error 'email->display-name "unexpected result"
+                             "result" actual
+                             "expected" expected))
+    ;; check no contract violations
+    (display-name-email actual)
+    (display-name-obfuscated-email actual)
+    (display-name-tags actual)
+    (for ([? '(#t #f)])
+      (display-name->preferred-tag actual #:obfuscate? ?)
+      (display-name->xexpr actual #:obfuscate? ?))))
+
+(define (display-name-tags dn)
+  (list (display-name-plain-tag dn)
+        (display-name-obfuscated-tag dn)))
+
+(define (display-name-obfuscated-email dn)
+  (match-define (struct* display-name ([local-part local-part]
+                                       [short-hash short-hash]))
+    dn)
+  (string-append local-part "λ" short-hash))
+
+(define (display-name->preferred-tag dn #:obfuscate? obfuscate?)
+  (if obfuscate?
+      (display-name-obfuscated-tag dn)
+      (display-name-plain-tag dn)))
+
+(define (display-name->xexpr dn  #:obfuscate? obfuscate?)
+  (match-define (struct* display-name ([email email]
+                                       [local-part local-part]
+                                       [short-hash short-hash]))
+    dn)
+  (cond
+    [obfuscate?
+     `(span ()
+            ,local-part
+            (span ([title "This author’s email address has been obfuscated."]
+                   [style "cursor: help;"]
+                   [class "text-muted"])
+                  (b ([style "font: monospace;"])
+                     "λ")
+                  ,short-hash))]
```

Comment:
In the light of a new day, it occurred to me that the `text-muted` CSS class is specific to the package website. I wanted to encapsulate details here, but I think the rendering details really ought to go in the `racket-pkg-website` repository so they design be iterated without having to change the core implementation. Instead, I think I should just expose `display-name-local-part` and `display-name-short-hash`&mdash;and at that point, I'd probably expose `display-name-plain-tag` and `display-name-obfuscated-tag`, too, and leave it for a the client code that wants them to implement `#:obfuscate?`. (I have gone back and forth on how much to provide more than once.)

Reply:
I've pushed this change.